lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs

```diff
@@ -529,19 +545,82 @@ getLegacyTransaction tid = do
     let inps = mempty
     pure $ Tx tid inps outs
 
+
+putWitness :: TxWitness -> Put
+putWitness (TxWitness bytes) = putByteString bytes
+
 putSignedTx :: [(TxIn, Coin)] -> [TxOut] -> [TxWitness] -> Put
 putSignedTx inputs outputs witnesses = do
-    putTx inputs outputs
+    putInputsOutputs inputs outputs
     unless (length inputs == length witnesses) $
         fail "number of witnesses must equal number of inputs"
     mapM_ putWitness witnesses
-  where
-    -- Assumes the `TxWitness` has been faithfully constructed
-    putWitness :: TxWitness -> Put
-    putWitness (TxWitness bytes) = putByteString bytes
 
-putTx :: [(TxIn, Coin)] -> [TxOut] -> Put
-putTx inputs outputs = do
+-- | Constructs a "normal" tx
+constructTransactionFragment
+    :: Monad m
+    => [(TxIn, TxOut)]
+    -> [TxOut]
+    -> (Address -> m (Hash "sigData" -> TxWitness))
+    -> m (Hash "Tx", SealedTx)
+constructTransactionFragment ins outs mkWit = do
+    constructGenericSignedTx FragmentTransaction
+        mempty
+        ins outs mkWit
+        (const mempty)
+
+-- | Constructs a generic transaction of the form
+--
+-- 1. Payload
+-- 2. Inputs / outputs
+-- 3. Witnesses (including signed hash of 1 and 2)
+-- 4. Payload authentication (calculated using 1,2,3 as input)
+constructGenericSignedTx
+    :: Monad m
+    => FragmentSpec
+    -- ^ The kind of transaction
+    -> Put
+    -- ^ Payload (empty for normal utxo transactions)
+    -> [(TxIn, TxOut)]
+    -> [TxOut]
+    -> (Address -> m (Hash "sigData" -> TxWitness))
+    -- ^ Function for producing witnesses for each address in the inputs.
+    -- E.g by looking up the private key and using it for signing with
+    -- @constructUtxoWitness@
+    -> (ByteString -> ByteString)
+    -- ^ Payload auth signing
```

Comment:
> Won't this always be `sign`?

I actually just found out that there is a [header byte in the account signatures](https://github.com/input-output-hk/chain-libs/blob/master/chain-impl-mockchain/doc/format.abnf#L168).

I don't know how other auth-payloads look. I'd be devastated if we at some point added some kind of extra "auth-prefix" argument to this function.

I think this is an elegant, future-proof (hopefully) way to go about it.

We can keep `constructGenericSignedTx` internal to the module though.

Reply:
Hmmm. But the function producing the signature is still `sign`. It is however true that the authenticated payload comes with a prefix. But that's independent of the signing process.

Reply:
> I'd be devastated if we at some point added some kind of extra "auth-prefix" argument to this function.

No! This is the point of refactoring and using an iterative approach. Avoid premature abstraction trying to think about "what the future might be". Either: 

- Get the answer (because we have it from the abnf format already)
- Do what you need to do to get today's job done

Reply:
Ok. We could hard-code the single account signature byte inside `constructGenericSignedTx`. We'd then get todays job done. Surely that can't be preferable?

I don't know which fragments we'd be interested in implementing. But pool registration transactions also have a header to the signature.

Would you rather that `constructGenericSignedTx` takes an argument of the type:
```haskell
data FragmentAuthentication
     = SignatureWithPrefix Put XPrv 
     | NoAuthentication
```
?